### PR TITLE
[Form] accept floats for input="string" in NumberType

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 4.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #31905, #32124
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Review by yceruto — APPROVED** [review]
Thanks!!

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/StringToFloatTransformer.php [hunk 1]
```diff
@@ -34,7 +34,7 @@ public function transform($value)
             return null;
         }
 
-        if (!\is_string($value) || !is_numeric($value)) {
+        if ((!\is_string($value) || !is_numeric($value)) && !\is_float($value)) {
             throw new TransformationFailedException('Expected a numeric string.');
```
Comment (yceruto) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/StringToFloatTransformer.php]: Maybe the message can also be improved: `'Expected a float or a numeric string.'`?

**Comment by fancyweb** [comment]
Shouldn't we throw a deprecation here so users actually uses string instead of floats in their entities?

**Comment by xabbuh** [comment]
I have updated the code a bit. The changes made for the `StringToFloatTransformer` have been reverted. Instead the `NumberType` takes care of the type cast.

This allows us to deprecate the behaviour with a better message in 4.4 as suggested in https://github.com/symfony/symfony/pull/32125#issuecomment-504317999 and we do not need to weaken the `StringToFloatTransformer`.

File: src/Symfony/Component/Form/Extension/Core/Type/NumberType.php [hunk 2]
```diff
@@ -37,6 +38,14 @@ public function buildForm(FormBuilderInterface $builder, array $options)
 
         if ('string' === $options['input']) {
             $builder->addModelTransformer(new StringToFloatTransformer($options['scale']));
+            $builder->addModelTransformer(new CallbackTransformer(
+                function ($value) {
+                    return \is_float($value) ? (string) $value : $value;
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/NumberType.php]: Should it accept an int the same way?

Comment (xabbuh) [reply src/Symfony/Component/Form/Extension/Core/Type/NumberType.php]: make sense

**Comment by xabbuh** [comment]
@fancyweb see #32130

**Review by yceruto — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Pull request merged**